src/pudl/glue/epacamd_eia.py

```diff
@@ -138,5 +143,400 @@ def transform(
             on=["plant_id_eia", "boiler_id"],
             how="inner",
         )
-
+    # TODO: Add mannual crosswalk
     return {"epacamd_eia": crosswalk_clean}
+
+
+##################
+# Add Sub-Plant ID
+##################
+
+
+# @asset
+def identify_subplants_epacamd_eia(
+    crosswalk_clean: pd.DataFrame,
+    generators_eia860: pd.DataFrame,
+    years=None,
+    states=None,
+) -> pd.DataFrame:
+    """Identify sub-plant IDs, GTN regressions, and GTN ratios.
+
+    TODO: convert to use dagster directly... I believe i need to convert transform above
+    into an asset that returns just crosswalk_clean (not {"epacamd_eia": crosswalk_clean})
+    """
+    cems_ids = unique_plant_ids_epacems(years, states)
+    # add subplant ids to the data
+    crosswalk_w_subplant_ids = generate_subplant_ids(
+        cems_ids, crosswalk_clean, generators_eia860
+    )
+    return crosswalk_w_subplant_ids
+
+
+def unique_plant_ids_epacems(
+    epacems_path: Path | None = None,
+    years: list[int] | None = None,
+    states: list[str] | None = None,
+) -> pd.DataFrame:
+    """Make unique annual plant_id_eia and emissions_unit_id_epa.
+
+    TODO: grab cems directly from its asset?
+
+    Returns:
+        dataframe with unique set of: "plant_id_eia", "report_year" and
+        "emissions_unit_id_epa"
+    """
+    logger.info("loading CEMS ids")
+    if epacems_path is None:
+        pudl_settings = pudl.workspace.setup.get_defaults()
+        epacems_path = (
+            Path(pudl_settings["parquet_dir"]) / "hourly_emissions_epacems.parquet"
+        )
+    annual_id_cols = ["plant_id_eia", "year", "emissions_unit_id_epa"]
+    epacems_dd = dd.read_parquet(
+        epacems_path,
+        use_nullable_dtypes=True,
+        engine="pyarrow",
+        index=False,
+        split_row_groups=True,
+        filters=year_state_filter(years, states),
+        columns=annual_id_cols,
+    )
+    epacems_ids = (
+        epacems_dd.compute().drop_duplicates().rename(columns={"year": "report_year"})
```

Comment:
Instead of reading from the epacems path, you can depend on the cems asset, which will load the cems data as a dask dataframe using the `PandasParquetIOManager.load_input()`. However, to depend on cems asset, the function you're working on must also be an asset. Where does this work fit into the DAG? Is it an analysis table or a part of the ETL glue?

Reply:
Can it be a function that is called within an asset? Although I suppose I could make this little guy an asset on its own.

Reply:
also @bendnorman ... how would i actually call `PandasParquetIOManager.load_input()`? is there a good example of how I can pass around `context`?

Reply:
Sure! It can be it's own function called in an asset. The easiest way to load assets using io managers is to just specify the asset name in the new asset's signature:

```python
@asset(io_manager_key="pudl_sqlite_io_manager")
def a():
   ...

@asset(io_manager_key="pudl_sqlite_io_manager")
def b(a): # will load asset "a" using the pudl_sqlite_io_manager.
   ...
```